engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -328,25 +337,59 @@ private void createInnerInstance(
         context, multiInstanceBody.getInnerActivity());
   }
 
-  private DirectBuffer wrapLoopCounter(final int loopCounter) {
-    loopCounterWriter.wrap(loopCounterVariableBuffer, 0);
+  private DirectBuffer wrapVariable(
+      final MutableDirectBuffer variableBuffer, final DirectBuffer variableView, final long value) {
+    variableWriter.wrap(variableBuffer, 0);
 
-    loopCounterWriter.writeInteger(loopCounter);
-    final var length = loopCounterWriter.getOffset();
+    variableWriter.writeInteger(value);
+    final var length = variableWriter.getOffset();
 
-    loopCounterVariableView.wrap(loopCounterVariableBuffer, 0, length);
-    return loopCounterVariableView;
+    variableView.wrap(variableBuffer, 0, length);
+    return variableView;
   }
 
   private Either<Failure, Boolean> satisfiesCompletionCondition(
       final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
     final Optional<Expression> completionCondition =
         element.getLoopCharacteristics().getCompletionCondition();
 
+    final ExpressionProcessor primaryContextExpressionProcessor = expressionBehavior.withPrimaryContext(
+        (variableName -> getVariable(context.getFlowScopeKey(), variableName)));
     if (completionCondition.isPresent()) {
-      return expressionBehavior.evaluateBooleanExpression(
+      return primaryContextExpressionProcessor.evaluateBooleanExpression(
           completionCondition.get(), context.getElementInstanceKey());
     }
     return Either.right(false);
   }
+
+  private DirectBuffer getVariable(
+      final long elementInstanceKey,
+      final String variableName) {
+    return switch (variableName) {
+      case "numberOfInstances" -> getNumberOfInstancesVariable(elementInstanceKey);
+
+      case "numberOfActiveInstances" -> getNumberOfActiveInstancesVariable(elementInstanceKey);
+
+      default -> null;
+    };
+  }
+
+  private DirectBuffer getNumberOfInstancesVariable(final long elementInstanceKey) {
+    return wrapVariable(
+        numberOfInstancesVariableBuffer,
+        numberOfInstancesVariableView,
+        stateBehavior.getElementInstance(elementInstanceKey).getNumberOfElementInstances());
+  }
+
+  private DirectBuffer getNumberOfActiveInstancesVariable(final long elementInstanceKey) {
+    final int numberOfActiveInstances =
+        stateBehavior
+                .getElementInstance(elementInstanceKey)
+                .getNumberOfActiveElementInstances()
+            - 1;
```

Comment:
🔧 Please add a comment on why this variable has to be lowered by 1. I think it's because the `getNumberOfActiveInstancesVariable` method is called while the child instance is completing but the active element instances value has not yet been decremented.